### PR TITLE
mongo dbstats charts

### PR DESCRIPTION
**Pull request by georgeok**
Implementation of dbstats

### CODE REVIEW

File: modules/mongodb/collect.go [hunk 1]
```diff
@@ -75,3 +75,32 @@ func (m *Mongo) metricExists(iface interface{}, chart *module.Chart) {
...
+	m.databases = databases
+
+	ms := map[string]int64{}
+	for _, database := range databases {
+		if !m.databasesMatcher.MatchString(database) {
+			continue
+		}
+		stats, err := m.mongoCollector.dbStats(database)
+		if err != nil {
+			return map[string]int64{}
```
Comment (ilyam8) [inline modules/mongodb/collect.go:101]:
```suggestion
		stats, err := m.mongoCollector.dbStats(database)
		if err != nil {
			return nil
```

And we need to log the error

File: modules/mongodb/mongodb.go [hunk 2]
```diff
@@ -44,6 +55,15 @@ func (m *Mongo) Init() bool {
 		return false
 	}
 
+	if !m.Databases.Empty() {
+		mMatcher, err := m.Databases.Parse()
+		if err != nil {
+			m.Errorf("error on creating 'databases' matcher : %v", err)
+			return false
+		}
+		m.databasesMatcher = matcher.WithCache(mMatcher)
```
Comment (ilyam8) [inline modules/mongodb/mongodb.go:64]:
```suggestion
		m.databasesMatcher = mMatcher
```

I don't think we need cache here, it should be pretty fast w/o it.

File: modules/mongodb/mongodb.go [hunk 3]
```diff
@@ -67,7 +87,16 @@ func (m *Mongo) Collect() map[string]int64 {
...
-	ms := m.serverStatusCollect()
+	ms := map[string]int64{}
+	for _, result := range []map[string]int64{
+		m.serverStatusCollect(),
+		m.dbStatsCollect(),
+	} {
+		for k, v := range result {
+			ms[k] = v
+		}
+	}
```
Comment (ilyam8) [inline modules/mongodb/mongodb.go:98]: I think we need to refactor it to

```go
if err := m.serverStatusCollect(ms); err != nil {
    ....
}
if err := m.dbStatsCollect(ms); err != nil {
    ....
}
```

We don't want to call "dbStatsCollect" if "serverStatusCollect" fails (mongodb is not responding).

Also, I think we need to name functions to "collectThis"/"collectThat".

File: modules/mongodb/mongodb.go [hunk 4]
```diff
@@ -83,5 +112,14 @@ func (m *Mongo) Cleanup() {
 }
 
 func (m *Mongo) initCharts() (*module.Charts, error) {
```
Comment (ilyam8) [inline modules/mongodb/mongodb.go:114]:
```suggestion
func (m Mongo) initCharts() (*module.Charts, error) {
```

I see we don't manipulate the receiver value.

File: modules/mongodb/mongodb.go [hunk 5]
```diff
@@ -83,5 +112,14 @@ func (m *Mongo) Cleanup() {
...
-	return &serverStatusCharts, nil
+	charts := module.Charts{}
```
Comment (ilyam8) [inline modules/mongodb/mongodb.go:115]:
```suggestion
	var charts module.Charts
```

Zero value will do

**Comment by ilyam8** [comment]
That is the menu, the order is:
 - default charts
 - database stats charts
 - optional charts

Screenshot 2021-08-16 at 20 11 48

Likely we will need to change the order. Not sure to which one, just highlighting it for now.

File: modules/mongodb/charts.go [hunk 6]
```diff
@@ -432,3 +441,113 @@ var (
...
+
+var (
+	dbStatsCollectionsChart = &module.Chart{
+		ID:    "collections",
+		Title: "Collections",
+		Units: "collections",
+		Fam:   "database_statistics",
+		Ctx:   "mongodb.collections",
+		Type:  module.Stacked,
+	}
```
[the remark this answers is not in the document]
Comment (georgeok) [reply modules/mongodb/charts.go]: done

File: modules/mongodb/charts.go [hunk 7]
```diff
@@ -432,3 +441,113 @@ var (
...
+	dbStatsViewsChart = &module.Chart{
+		ID:    "views",
+		Title: "Views",
+		Units: "views",
+		Fam:   "database_statistics",
+		Ctx:   "mongodb.views",
+		Type:  module.Stacked,
+	}
```
[the remark this answers is not in the document]
Comment (georgeok) [reply modules/mongodb/charts.go]: done

File: modules/mongodb/collect.go [hunk 8]
```diff
@@ -75,3 +75,32 @@ func (m *Mongo) metricExists(iface interface{}, chart *module.Chart) {
...
+
+// dbStatsCollect creates the map[string]int64 for the available dims
+// it calls listDatabase and then dbstats for each database internally
+func (m *Mongo) dbStatsCollect() map[string]int64 {
+	if m.databasesMatcher == nil {
+		return map[string]int64{}
+	}
+	databases, err := m.mongoCollector.listDatabaseNames()
+	if err != nil {
+		return map[string]int64{}
```
[the remark this answers is not in the document]
Comment (georgeok) [reply modules/mongodb/collect.go]: done

File: modules/mongodb/charts.go [hunk 9]
```diff
@@ -432,3 +441,113 @@ var (
...
+		Units: "bytes",
+		Fam:   "database_statistics",
+		Ctx:   "mongodb.storage_size",
+		Type:  module.Stacked,
+	}
+)
+
+// dimsForDbStats adds dimensions for new databases and
+// removes for dropped
+func (m *Mongo) dimsForDbStats(newDatabases []string) {
```
[the remark this answers is not in the document]
Comment (georgeok) [reply modules/mongodb/charts.go]: done

File: modules/mongodb/mongodb.go (hunk 4 above)
Comment (georgeok) [reply modules/mongodb/mongodb.go]: done

File: modules/mongodb/mongodb.go (hunk 2 above)
Comment (georgeok) [reply modules/mongodb/mongodb.go]: I have removed the cache here

File: modules/mongodb/charts.go [hunk 10]
```diff
@@ -432,3 +441,113 @@ var (
...
+	if len(newDatabases) == 0 {
+		return
+	}
+	if m.databasesMatcher == nil {
+		return
+	}
```
[the remark this answers is not in the document]
Comment (georgeok) [reply modules/mongodb/charts.go]: Good suggestion, I've refactored the code a bit to use the matcher after `m.mongoCollector.listDatabaseNames()`

File: modules/mongodb/mongodb.go (hunk 3 above)
Comment (georgeok) [reply modules/mongodb/mongodb.go]: done

File: modules/mongodb/charts.go [hunk 11]
```diff
@@ -432,3 +441,113 @@ var (
...
+	// remove dims for not existing databases
+	m.removeObsoleteDims(newDatabases)
+
+	// add dimensions for new databases
+	for _, chart := range dbStatsCharts {
+		for _, name := range newDatabases {
+			if !m.databasesMatcher.MatchString(name) {
+				continue
+			}
+			if !m.dimsEnabled[chart.ID+"_"+name] {
```
[the remark this answers is not in the document]
Comment (georgeok) [reply modules/mongodb/charts.go]: Done it a bit differently though. We are not caching enabled dims anymore. `discoveredDBs` previous name was `databases` hold the names of the databases reported by the list command. This is also use when we want to remove dims for dropped databases

File: modules/mongodb/charts.go [hunk 12]
```diff
@@ -432,3 +441,109 @@ var (
...
+	if len(databases) == 0 {
+		return
+	}
+
+	// remove dims for not existing databases
+	m.removeObsoleteDims(databases)
+
+	// add dimensions for new databases
+	for _, database := range sliceDiff(databases, m.discoveredDBs) {
+		for _, chart := range dbStatsCharts {
```
Comment (ilyam8) [inline modules/mongodb/charts.go:504]: `dbStatsCharts` is a global variable, it is shared across different `mongodb` jobs. We shouldn't modify/use the values from `dbStatsCharts` slice.

File: modules/mongodb/charts.go [hunk 13]
```diff
@@ -432,3 +441,109 @@ var (
...
+	// update the cache
+	m.discoveredDBs = databases
+}
+
+// removeObsoleteDims removes dimensions from dbstats
+// charts for dropped databases
+func (m *Mongo) removeObsoleteDims(newDatabases []string) {
+	diff := sliceDiff(m.discoveredDBs, newDatabases)
+	for _, name := range diff {
+		for _, chart := range dbStatsCharts {
```
Comment (ilyam8) [inline modules/mongodb/charts.go:524]: `dbStatsCharts` is a global variable, it is shared across different `mongodb` jobs. We shouldn't modify/use the values from `dbStatsCharts` slice.

File: modules/mongodb/model.go [hunk 14]
```diff
@@ -233,3 +233,23 @@ type WiredTiger struct {
...
 }
+
+type dbStats struct {
+	Collections int64 `bson:"collections" stm:"collections"`
+	Views       int64 `bson:"views" stm:"views"`
+	Indexes     int64 `bson:"indexes" stm:"indexes"`
+	Objects     int64 `bson:"objects" stm:"documents"`
+	DataSize    int64 `bson:"dataSize" stm:"data_size"`
+	IndexSize   int64 `bson:"indexSize" stm:"index_size"`
+	StorageSize int64 `bson:"storageSize" stm:"storage_size"`
```
Comment (ilyam8) [inline modules/mongodb/model.go:244]:
```suggestion
	Collections int64 `bson:"collections"`
	Views       int64 `bson:"views"`
	Indexes     int64 `bson:"indexes"`
	Objects     int64 `bson:"objects"`
	DataSize    int64 `bson:"dataSize"`
	IndexSize   int64 `bson:"indexSize"`
	StorageSize int64 `bson:"storageSize"`
```

We don't `stm.ToMap` this struct, the tags can be removed

File: modules/mongodb/mongodb.go [hunk 15]
```diff
@@ -4,11 +4,13 @@ import (
...
 	"github.com/netdata/go.d.plugin/agent/module"
+	"github.com/netdata/go.d.plugin/pkg/matcher"
 )
 
 type Config struct {
-	Uri     string        `yaml:"uri"`
-	Timeout time.Duration `yaml:"timeout"`
+	Uri       string             `yaml:"uri"`
+	Timeout   time.Duration      `yaml:"timeout"`
+	Databases matcher.SimpleExpr `yaml:"discoveredDBs"`
```
Comment (ilyam8) [inline modules/mongodb/mongodb.go:13]:
```suggestion
	Databases matcher.SimpleExpr `yaml:"databases"`
```

File: modules/mongodb/charts.go [hunk 16]
```diff
@@ -432,3 +441,109 @@ var (
...
+		Type:  module.Stacked,
+	}
+)
+
+// updateDBStatsCharts adds dimensions for new databases and
+// removes for dropped
+func (m *Mongo) updateDBStatsCharts(databases []string) {
+	if len(databases) == 0 {
+		return
+	}
```
Comment (ilyam8) [inline modules/mongodb/charts.go:497]: I think we shouldn't early return here.

- 4 databases in total
- 4 => filtering => 1
- we add 1 db
- a user removes that db
- 3 databases in total
- 3 => filtering => 0
- we don't remove the db

Comment (georgeok) [reply modules/mongodb/charts.go]: good point, fixed

File: modules/mongodb/model.go (hunk 14 above)
Comment (georgeok) [reply modules/mongodb/model.go]: done

File: modules/mongodb/charts.go (hunk 12 above)
Comment (georgeok) [reply modules/mongodb/charts.go]: change dbStatsCharts to a state variable. Nice catch again

File: modules/mongodb/mongodb.go [hunk 17]
```diff
@@ -4,11 +4,13 @@ import (
...
 	"github.com/netdata/go.d.plugin/agent/module"
+	"github.com/netdata/go.d.plugin/pkg/matcher"
 )
 
 type Config struct {
-	Uri     string        `yaml:"uri"`
-	Timeout time.Duration `yaml:"timeout"`
+	Uri       string             `yaml:"uri"`
+	Timeout   time.Duration      `yaml:"timeout"`
+	Databases matcher.SimpleExpr `yaml:"databasesupdateDBStatsCharts"`
```
Comment (ilyam8) [inline modules/mongodb/mongodb.go:13]:
```suggestion
	Databases matcher.SimpleExpr `yaml:"databases"`
```

File: modules/mongodb/mongodb.go [hunk 18]
```diff
@@ -4,11 +4,13 @@ import (
 	"time"
 
 	"github.com/netdata/go.d.plugin/agent/module"
+	"github.com/netdata/go.d.plugin/pkg/matcher"
 )
 
 type Config struct {
-	Uri     string        `yaml:"uri"`
-	Timeout time.Duration `yaml:"timeout"`
+	Uri       string             `yaml:"uri"`
```
Comment (ilyam8) [inline modules/mongodb/mongodb.go:11]: Let's make it URI

File: modules/mongodb/mongodb.go (hunk 17 above)
Comment (georgeok) [reply modules/mongodb/mongodb.go:13]: sorry, that was an accident

File: modules/mongodb/mongodb.go (hunk 18 above)
Comment (georgeok) [reply modules/mongodb/mongodb.go]: done

File: modules/mongodb/charts.go [hunk 19]
```diff
@@ -432,3 +441,105 @@ var (
...
+// removeObsoleteDims removes dimensions from dbstats
+// charts for dropped databases
+func (m *Mongo) removeObsoleteDims(newDatabases []string) {
+	diff := sliceDiff(m.discoveredDBs, newDatabases)
+	for _, name := range diff {
+		for _, chart := range *m.chartsDbStats {
+			id := chart.ID + "_" + name
+			err := chart.MarkDimRemove(id, true)
+			if err != nil {
+				m.Warningf("failed to remove dimension %s with error: %s", id, err.Error())
```
Comment (ilyam8) [inline modules/mongodb/charts.go:524]:
```suggestion
			if err != nil {
				m.Warningf("failed to remove dimension %s with error: %v", id, err)
```

"Error()" is called for both "v" and "s" verbs

File: modules/mongodb/charts.go [hunk 20]
```diff
@@ -432,3 +441,105 @@ var (
...
+	// remove dims for not existing databases
+	m.removeObsoleteDims(databases)
+
+	// add dimensions for new databases
+	for _, database := range sliceDiff(databases, m.discoveredDBs) {
+		for _, chart := range *m.chartsDbStats {
+			id := chart.ID + "_" + database
+			err := chart.AddDim(&module.Dim{ID: id, Name: database, Algo: module.Absolute})
+			if err != nil {
+				m.Errorf("failed to add dim: %s, %s", id, err)
```
[the remark this answers is not in the document]
Comment (georgeok) [reply modules/mongodb/charts.go]: done

File: modules/mongodb/charts.go (hunk 19 above)
Comment (georgeok) [reply modules/mongodb/charts.go]: done

File: modules/mongodb/charts.go [hunk 21]
```diff
@@ -432,3 +441,105 @@ var (
...
+		Ctx:   "mongodb.database_storage_size",
+		Type:  module.Stacked,
+	}
+)
+
+// updateDBStatsCharts adds dimensions for new databases and
+// removes for dropped
+func (m *Mongo) updateDBStatsCharts(databases []string) {
+	// remove dims for not existing databases
+	m.removeObsoleteDims(databases)
```
[the remark this answers is not in the document]
Comment (georgeok) [reply modules/mongodb/charts.go]: I picked the first suggestion!

**Review by ilyam8 — APPROVED** [review]

**Comment by ilyam8** [comment]
@georgeok, many thanks! Really nice work 👍

**Pull request merged**